Pack Command to Create a Buildpack Repo

jkutner opened the pull request:
[Readable](https://github.com/buildpacks/rfcs/blob/pack-buildpack-create/text/0000-pack-buildpack-create.md)

ryanmoran commented:
A Paketo-specific and opinionated example: https://github.com/paketo-community/bootstrapper. May add some evidence to the utility of such a tool.

jkutner commented:
@ryanmoran thanks. I added that to the prior art section

dfreilich commented on text/0000-pack-buildpack-create.md:
```diff
@@ -0,0 +1,74 @@
+[alternatives]: #alternatives
+
+- Do it by hand
+- A separate tool (no Pack) can create the scaffolding
```
```suggestion
- A separate tool (not Pack) can create the scaffolding
```
This may tie in to the plugin for pack discussion

dfreilich commented on text/0000-pack-buildpack-create.md:
````diff
@@ -0,0 +1,74 @@
+```
+$ pack buildpack create <id> \
+    --language <lang> \
+    --stack <stack>
````
I'd like users to be able to provide multiple stacks

jkutner replied:
yea, that was my intention here (like the `--buildpack` flag). I'll call that out.

dfreilich reviewed (approved):
Overall, I definitely like the idea. As I mentioned in the PR to pack, it was an issue discussed in pack over two years ago (https://github.com/buildpacks/pack/issues/82), and it's definitely could add value to the project by making it clearer how to get started creating buildpacks.

dfreilich commented on text/0000-pack-buildpack-create.md:
```diff
@@ -0,0 +1,80 @@
+
+* `language` - a choice of programming languages to use for the template. Pack will support Bash and Go at a minimum.
+* `stack` - a set of compatible stacks for the buildpack. may be specificied multiple times.
+* `path` - the location on the filesystem to generate the artifacts
```
The default path would be the current directory?

hone replied:
I assume so, but it'd be good to specify.

```
$ pack buildpack create <foo>
```
would make a directory `foo/` in the current working dir.

jkutner replied:
@hone agree, I wasn't sure how to handle this because the id `foo/bar` shouldn't create two directories (`./foo/bar/`), and I thought `foo_bar` was kind of ugly. I noticed that the Paketo Bootstrapper would create a one directory: `bar/`. I like that

hone replied:
good call, forgot about the namespace bit.

hone commented on text/0000-pack-buildpack-create.md:
````diff
@@ -0,0 +1,80 @@
+```
+$ pack buildpack create <id> \
+    --language <lang> \
+    --stack <stack>
````
I believe this flag is missing here.

```suggestion
    --stack <stack>
    --path <path>
```

hone commented on text/0000-pack-buildpack-create.md:
```diff
@@ -0,0 +1,80 @@
+
+Where the flags include:
+
+* `language` - a choice of programming languages to use for the template. Pack will support Bash and Go at a minimum.
```
How does one get a language in this list? Will we support 3rd party at all?

micahyoung replied:
+1 on defaulting to `bash`, even when `pack` runs on Windows. 

Most Windows desktop users today would likely be building Linux buildpacks (on LCOW daemon). If they want to create a buildpack for WCOW, right now, there's not a clear user-story for the rest of the experience given there's currently not a beginner-friendly ecosystem of Windows stacks/builders anyways. They would be likely be better served by a separate samples/template/tutorial walk-through for the time being.

jkutner replied:
Thanks @micahyoung!

We also discussed in the working group that in the future we may add a `--template` option, and curate a repository of templates for different languages. Contributors could PR into the repo to add support for other langs (or even for things like packit).

samj1912 replied:
This is very interesting! An alternative to consider would be having a functionality similar to [cookiecutter](https://github.com/cookiecutter/cookiecutter) where you can provide a URI to a template (either a VCS uri or on the local fs). This would make it easier to handle new template contributions and they don't need to be put in a single repository.

elbandito commented on text/0000-pack-buildpack-create.md:
```diff
@@ -0,0 +1,81 @@
+# Definitions
+[definitions]: #definitions
+
+- *Scaffold* - the essential files and directory required for a buildpack
```
@jkutner did you mean `directories`?

jkutner replied:
addressed in edf3764

jabrown85 commented on text/0000-pack-buildpack-create.md:
````diff
@@ -0,0 +1,81 @@
+```
+$ pack buildpack create <id> \
+    --language <lang> \
+    --stack <stack>
````
```suggestion
    --stack <stack> \
```

jkutner replied:
addressed in edf3764

jkutner commented on text/0000-pack-buildpack-create.md:
````diff
@@ -0,0 +1,81 @@
+Add a new command to Pack:
+
+```
+$ pack buildpack create <id> \
````
```suggestion
$ pack buildpack new <id> \
```

jkutner replied:
addressed in edf3764

natalieparellano commented on text/0000-pack-buildpack-create.md:
```diff
@@ -0,0 +1,81 @@
+- [`rails new`](https://guides.rubyonrails.org/command_line.html)
+- [`npm init`](https://docs.npmjs.com/cli/v6/commands/npm-init)
+
+# Unresolved Questions
```
Is there any mechanism to ensure that the scaffold stays up to date when apis change?

jkutner replied:
No, I would like to keep that out-of-scope

samj1912 replied:
Again - something to consider in the future is a functionality similar to https://github.com/cruft/cruft/  (see my cookiecutter template related comment above). This allows users to keep their repositories up to date with the template. It essentially performs a 3 way merge.

jabrown85 reviewed (approved):
I think this is a great start. I'm sure we'll expand on this in the future, but this goes a long way to getting someone off the ground.

jkutner commented:
TODO from Working Group:
* change the name of the command to `new`
* Remove the `--language` flag (default to bash, and support templates later, and put in future work)
* Add buildpack type (stack, meta, etc) to future work
* Ping @micahyoung to ask about Windows opinions

jkutner commented:
Updated the proposal based on feedback from working group, include:

* changed the name of the command to `pack buildpack new`
* removed the `--language` flag in favor of defaulting to bash

dfreilich commented on text/0000-pack-buildpack-create.md:
```diff
@@ -0,0 +1,99 @@
+* `bin/build`
+* `bin/detect`
+
+The `buildpack.toml` will be pre-poluated with the values provided in the CLI command and some defaults (like `0.0.0` for the version).
```
Would it make sense to do `0.0.1` as the default version, rather than `0.0.0`?

jkutner replied:
i don't have a strong opinion

dfreilich reviewed (approved):
LGTM

jkutner commented on text/0000-pack-buildpack-create.md:
````diff
@@ -0,0 +1,81 @@
+```
+$ pack buildpack create <id> \
+    --language <lang> \
+    --stack <stack>
````
I'm fine with adding those. `stack` is the only one I thought was weird to default (but I think did anyways)

jkutner replied:
Addressed in 559bf47

jkutner commented on text/0000-pack-buildpack-create.md:
````diff
@@ -0,0 +1,81 @@
+Add a new command to Pack:
+
+```
+$ pack buildpack create <id> \
````
@ekcasey not in this proposal.

jkutner commented:
@nebhale @sclevine PTAL

sclevine commented:
FCP starting 2/10

samj1912 commented on text/0000-pack-buildpack-create.md:
```diff
@@ -0,0 +1,105 @@
+
+## Future Work
+
+In the future, we may add support for a `--template` flag, which would create buildpacks from templates stored in a curated repository. In this way we can support many different languages, and even support versions the same language with different opinions.
```
This is very interesting! An alternative to consider would be having a functionality similar to [cookiecutter](https://github.com/cookiecutter/cookiecutter) where you can provide a URI to a template (either a VCS uri or on the local fs). This would make it easier to handle new template contributions and they don't need to be put in a single repository (which may be a bottleneck in terms of reviews/flexibility?).
